vidyut-lipi/src/lib.rs

```diff
@@ -1,118 +1,551 @@
+}
+
+
+// transliterate instr from inmap to outmap, and add to string.
+// inmap - input translation map.
+// outmap : output translation map.
+// outstr: output string.
+// prev: Indicates previous token type.
+
+
+pub fn transliterate(instr: String, inmap: &TranslationMap,
+		     outmap: &TranslationMap, outstr: &mut String) {
+
+    let wvec: Vec<char> = instr.chars().collect();
+    let count: usize = instr.chars().count();
+    let mut sindx = 0;
+    let mut prevch: u8 = TOK_UNKNOWN_TYPE;
+    let mut keystr: String = "".to_string();
+    let mut encoding_on: bool  = true;
+    let mut mnode: &Node;
+
+    while sindx < count {
+	// push a character into lookup string.
+	// (1)
+	let curch = wvec[sindx];
+//	println!("ch: {} indx: {}", curch, sindx);
+	sindx = sindx + 1;
+	keystr.push(curch);
+	let node = inmap.get_enc_map(keystr.to_string());
+	if node.is_none() {
+	    // no match was found for the current string. 
+	    // If length of keystr is greater than 1, then
+	    // we should have a match at keystr[:-1]
+	    if keystr.len() > 1 {
+		keystr.pop();
+		// Move the character that did not match back into
+		// input stream. 
+		sindx = sindx - 1;
+	    }
+	    let enode = inmap.get_enc_map(keystr.to_string());
+	    // this should match else we have a problem. 
+	    if enode.is_some() {
+	     
+		let mnode = enode.unwrap();
+		if mnode.tok_val.eq("TOK_ENCODING_ESCAPE")
+		    && mnode.tok_type == TOK_ENCODING_ESCAPE
+		{
+		    encoding_on = !encoding_on;
+		} else {
+		    if encoding_on {
+			outmap.update_string_with_token_with_implicit(
+			    mnode.tok_val.clone(),
+			    outstr, mnode.implicit, prevch);
+			prevch = mnode.tok_type;
+		    } else {
+			outstr.push_str(keystr.as_str());
+			keystr.clear();			
+		    }
+		}
+		keystr.clear();
+	    } else {
+		// this is not a mapping string.
+		// Copy input to output.
+		outstr.push_str(keystr.as_str());
+		keystr.clear();
+	    }
+	} else {
+	    // there is a match.
+	    mnode = node.unwrap();
+	    if mnode.suffix == false {
+		if mnode.tok_val.eq("TOK_ENCODING_ESCAPE") &&
+		    mnode.tok_type == TOK_ENCODING_ESCAPE
+		{
+		    encoding_on = !encoding_on;
+		    keystr.clear();
+		} else {
+		    if encoding_on {
+			outmap.update_string_with_token_with_implicit(
+			    mnode.tok_val.clone(),
+			    outstr, mnode.implicit, prevch);
+			prevch = mnode.tok_type;
+			keystr.clear();
+		    } else {
+			outstr.push_str(keystr.as_str());
+			keystr.clear();
+		    }
+		}
+	    }
+	}	
+    }
+    if !keystr.is_empty() {
+	let node = inmap.get_enc_map(keystr.to_string());
+	if node.is_some() && encoding_on {
+	    mnode = node.unwrap();
+	    outmap.update_string_with_token_with_implicit(mnode.tok_val.clone(),
+							  outstr,
+							  mnode.implicit,
+							  prevch);
+	    keystr.clear();
+	} else {
+	    outstr.push_str(keystr.as_str());
+	    keystr.clear();	    
+	}
+    }
+}
+
+
+fn check_for_prefix(keyvecs: &Vec<String>, b: &String) -> bool {
```

Comment:
```
fn has_prefix(keys: &[String], prefix: &str) -> bool {
  keys.iter().any(|k| k.starts_with(prefix))
}
```

Since this is a one-liner anyway, it could also replace the function entirely.

Reply:
ok